### PR TITLE
[SPARK-41173][SQL] Move `require()` out from the constructors of string expressions

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr aims to move `require()` out from the constructors of string expressions, include  `ConcatWs` and `FormatString`.
The args number checking logic moved into `checkInputDataTypes()`.

### Why are the changes needed?
Migration onto error classes unifies Spark SQL error messages.

### Does this PR introduce _any_ user-facing change?
Yes. The PR changes user-facing error messages.

### How was this patch tested?
Pass GitHub Actions

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/results/ansi/string-functions.sql.out [hunk 1]
```diff
@@ -14,7 +29,7 @@ select format_string()
 struct<>
 -- !query output
 org.apache.spark.sql.AnalysisException
-requirement failed: format_string() should take at least 1 argument; line 1 pos 7
+0; line 1 pos 7
```
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/results/ansi/string-functions.sql.out:32]: hmm, why is that? We should see similar error as for `select concat_ws()`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 2]
```diff
@@ -1662,8 +1675,7 @@ case class StringRPad(str: Expression, len: Expression, pad: Expression = Litera
 // scalastyle:on line.size.limit
 case class FormatString(children: Expression*) extends Expression with ImplicitCastInputTypes {
 
-  require(children.nonEmpty, s"$prettyName() should take at least 1 argument")
-  if (!SQLConf.get.getConf(SQLConf.ALLOW_ZERO_INDEX_IN_FORMAT_STRING)) {
+  if (children.nonEmpty && !SQLConf.get.getConf(SQLConf.ALLOW_ZERO_INDEX_IN_FORMAT_STRING)) {
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:1678]: My mistake, need to add nonEmpty condition

File: sql/core/src/test/resources/sql-tests/results/ansi/string-functions.sql.out (hunk 1 above)
Comment (LuciferYang) [reply sql/core/src/test/resources/sql-tests/results/ansi/string-functions.sql.out]: fixed

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @LuciferYang.

**Pull request closed without merging**